File: app/image_capture.py
import logging
import asyncio
from typing import Optional, Tuple

import cv2
import pyautogui
import numpy as np

from app.window_controroller import WindowController


class ImageCapture:
    def __init__(self, window_controller: WindowController):
        self.window_controller = window_controller

    async def capture_pair(
        self, delay: float = 0.1
    ) -> Tuple[Optional[np.ndarray], Optional[np.ndarray]]:
        region = self.window_controller.region
        if region is None:
            logging.warning("Capture failed: window not set.")
            return None, None

        try:
            img1 = await self._shot(region)
            await asyncio.sleep(delay)
            img2 = await self._shot(region)
        except Exception as e:
            logging.warning(f"Screenshot failed: {e}")
            return None, None

        return img1, img2

    async def _shot(self, region) -> np.ndarray:
        pil_img = await asyncio.to_thread(pyautogui.screenshot, None, region)
        cv_img = cv2.cvtColor(np.array(pil_img), cv2.COLOR_RGB2BGR)
        return cv_img

File: App/ImageCapture.cs
using System.Drawing;
using OpenCvSharp;
using OpenCvSharp.Extensions;

namespace DekapuAutoOpencv.Core;

public class ImageCapture(WindowController windowController)
{
    public async Task<(Mat? Prev, Mat? Curr)> CapturePairAsync(double delay = 0.1)
    {
        var region = windowController.Region;
        if (region is null)
        {
            Console.WriteLine("Capture failed: window not set.");
            return (null, null);
        }

        try
        {
            var img1 = await Task.Run(() => Shot(region.Value));
            await Task.Delay(TimeSpan.FromSeconds(delay));
            var img2 = await Task.Run(() => Shot(region.Value));
            return (img1, img2);
        }
        catch (Exception e)
        {
            Console.WriteLine($"Screenshot failed: {e.Message}");
            return (null, null);
        }
    }

    private static Mat Shot((int Left, int Top, int Width, int Height) region)
    {
        var (left, top, width, height) = region;

        using var bmp = new Bitmap(width, height, System.Drawing.Imaging.PixelFormat.Format32bppArgb);
        using var g = Graphics.FromImage(bmp);
        g.CopyFromScreen(left, top, 0, 0, new System.Drawing.Size(width, height));

        // BitmapConverter.ToMat returns CV_8UC4 (BGRA) for 32bpp bitmaps
        var mat = BitmapConverter.ToMat(bmp);

        // Convert BGRA → BGR to match Python's cv2.cvtColor(RGB → BGR) pipeline
        Cv2.CvtColor(mat, mat, ColorConversionCodes.BGRA2BGR);
        return mat;
    }
}
